lib/pavilion/plugins/commands/graph.py

```diff
@@ -0,0 +1,313 @@
+                                      label="average({})".format(label))
+
+        matplotlib.pyplot.ylabel(args.ylabel)
+        matplotlib.pyplot.xlabel(args.xlabel)
+        matplotlib.pyplot.legend()
+
+        matplotlib.pyplot.savefig(args.filename)
+        output.fprint("Completed. Graph saved as '{}.png'."
+                      .format(args.filename), color=output.GREEN,
+                      file=self.outfile)
+
+    def gather_results(self, evaluations, test_results) -> Dict:
+        """
+        Gather and format a test run objects results.
+        :param evaluations: The evaluations dictionary to be used to gather
+               results.
+        :param test_results: A test run's result dictionary.
+        :return: result, a dictionary containing parsed/formatted results for
+                 the given test run.
+        """
+
+        evaluate_results(test_results, evaluations)
+
+        # X value evaluations should only result in a list when graphing
+        # individual node results from a single test run.
+        if isinstance(test_results['x'], list):
+            if not isinstance(test_results['x'][-1], (int, str, float)):
+                raise ResultTypeError("x value evaluation '{}' resulted in a "
+                                      "list of invalid type {}."
+                                      .format(evaluations['x'],
+                                              type(test_results['x'][-1])
+                                              .__name__))
+            results = {}
+            for i in range(len(test_results['x'])):
+                x = test_results['x'][i]
+
+                # Determines if x value is node name, if so convert name to int.
+                if isinstance(x, str):
+                    node = re.match(r'[a-zA-Z]*(\d*)$', x)
+                    if node:
+                        x = int(node.groups()[0])
+
+                evals = {}
+                for key in evaluations.keys():
+                    if key == 'x':
+                        continue
+                    if not isinstance(test_results[key], list):
+                        raise ResultTypeError("y value evaluation '{}' resulted"
+                                              " in invalid type {}."
+                                              .format(evaluations[key],
+                                                      type(test_results[key])
+                                                      .__name__))
+                    evals.update({key: test_results[key][i]})
+
+                results[x] = evals
+
+        elif isinstance(test_results['x'], (int, float, str)):
+            results = {}
+            evals = {}
+
+            x = test_results['x']
+
+            for key in evaluations.keys():
+                if key == 'x':
+                    continue
+                evals.update({key: test_results[key]})
+            results[x] = evals
```

Comment:
Make this the first case, it will be much easier to understand what is going on.